### PR TITLE
Migrate access control

### PR DESCRIPTION
**Pull request by andrew-fleming**
Resolves (partial) #548.

Follows #584 with `#[internal]` macros for consistency.

✅ ~TODO: fix commented sections for ERC165 (after merging #582).~

Note: error messages do not include the `AccessControl:` prefix because they will exceed the 31 character limit.

#### PR Checklist

- [x] Tests
- [ ] Tried the feature on a public network
- [ ] Documentation

### CODE REVIEW

File: src/openzeppelin/tests/test_accesscontrol.cairo [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    setup();
+    set_caller_address(ACCOUNT2());
+    AccessControl::renounce_role(DEFAULT_ADMIN_ROLE, ACCOUNT1());
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_set_role_admin() {
+    setup();
+    AccessControl::_set_role_admin(DEFAULT_ADMIN_ROLE, SOME_OTHER_ROLE);
```
Comment (martriay) [inline src/openzeppelin/tests/test_accesscontrol.cairo:109]: these names are quite confusing. `SOME_OTHER_ROLE` is the admin role of `DEFAULT_ADMIN_ROLE`, yet the latter has "ADMIN" in it. this makes a bit hard to follow the tests.

File: src/openzeppelin/tests/test_accesscontrol.cairo [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    assert(
+        AccessControl::get_role_admin(DEFAULT_ADMIN_ROLE) == SOME_OTHER_ROLE,
+        'Should set the admin role'
+    );
+
+    // Test role admin cycle
+    AccessControl::grant_role(SOME_OTHER_ROLE, ACCOUNT2());
+    assert(AccessControl::has_role(SOME_OTHER_ROLE, ACCOUNT2()), 'ACCOUNT2 should have role');
```
Comment (martriay) [inline src/openzeppelin/tests/test_accesscontrol.cairo:117]: if i understand correctly, here we shouldn't be asserting since we're just preparing the test environment to check whether `_set_role_admin` is working properly. what we should be testing that accounts with the `SOME_OTHER_ROLE` role can grant `DEFAULT_ADMIN_ROLE`.

File: src/openzeppelin/tests/test_accesscontrol.cairo [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+    set_caller_address(ACCOUNT2());
+    AccessControl::revoke_role(DEFAULT_ADMIN_ROLE, ACCOUNT1());
+
+    assert(
+        !AccessControl::has_role(DEFAULT_ADMIN_ROLE, ACCOUNT1()), 'ACCOUNT1 should not have role'
```
Comment (martriay) [inline src/openzeppelin/tests/test_accesscontrol.cairo:123]:
```suggestion
        !AccessControl::has_role(DEFAULT_ADMIN_ROLE, ACCOUNT1()), 'ACCOUNT2 should revoke role'
```
here we're testing whether the new admin role can revoke role or not. in line with the above suggestions, we should stick with who's the admin and who's the account being granted/revoked so tests are easier to follow. we should rename both the roles and the accounts.

**Review by martriay — COMMENTED** [review]
nice job! left some comments and suggestions

**Review by martriay — COMMENTED** [review]
nice job! left some comments and suggestions

File: src/openzeppelin/tests/test_accesscontrol.cairo [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    AccessControl::renounce_role(DEFAULT_ADMIN_ROLE, ACCOUNT1());
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_set_role_admin() {
+    setup();
+    AccessControl::_set_role_admin(DEFAULT_ADMIN_ROLE, SOME_OTHER_ROLE);
+    assert(
+        AccessControl::get_role_admin(DEFAULT_ADMIN_ROLE) == SOME_OTHER_ROLE,
```
Comment (martriay) [inline src/openzeppelin/tests/test_accesscontrol.cairo:111]: there's no `get_role_admin` test. we could consider this is being tested in here, although i wonder whether we should test for other roles returning the default_admin_role as default

File: src/openzeppelin/tests/test_accesscontrol.cairo [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    set_caller_address(ACCOUNT2());
+    AccessControl::assert_only_role(DEFAULT_ADMIN_ROLE);
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_grant_role() {
+    setup();
+    AccessControl::grant_role(DEFAULT_ADMIN_ROLE, ACCOUNT2());
+    assert(AccessControl::has_role(DEFAULT_ADMIN_ROLE, ACCOUNT2()), 'Role should be granted');
```
Comment (martriay) [inline src/openzeppelin/tests/test_accesscontrol.cairo:56]: do we need to also test `has_role`? when it should return `true` and when `false`

File: src/openzeppelin/access/accesscontrol.cairo [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        fn has_role(role: felt252, account: ContractAddress) -> bool {
+            role_member::read((role, account))
+        }
+
+        fn get_role_admin(role: felt252) -> felt252 {
+            role_admin::read(role)
+        }
+
+        fn grant_role(role: felt252, account: ContractAddress) {
+            let admin: felt252 = get_role_admin(role);
```
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/access/accesscontrol.cairo]: Nope, force of habit

File: src/openzeppelin/access/accesscontrol.cairo [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+#[contract]
+mod AccessControl {
+    // OZ modules
+    use super::IAccessControl;
+    use openzeppelin::introspection::erc165::ERC165;
+
+    // Constants
+    use openzeppelin::utils::constants::DEFAULT_ADMIN_ROLE;
```
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/access/accesscontrol.cairo]: I mean, we are just using both constants once, so we can just import the constants mod and do `constants::<target>`. I went back and forth on this originally, but thinking about it now, it might be better that way

File: src/openzeppelin/access/accesscontrol.cairo [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    #[view]
+    fn supports_interface(interface_id: u32) -> bool {
+        ERC165::supports_interface(interface_id)
+    }
+
+    #[view]
+    fn has_role(role: felt252, account: ContractAddress) -> bool {
+        AccessControl::has_role(role, account)
```
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/access/accesscontrol.cairo:78]: Haha agreed. The impl includes an extra `AccessControl` e.g. impl: `AccessControl::AccessControl::has_role` vs mod: `AccessControl::has_role`. I'll add the `Impl` suffix (we should definitely apply this to the rest of the contracts)

File: src/openzeppelin/access/accesscontrol.cairo [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+            let caller: ContractAddress = get_caller_address();
+            assert(caller == account, 'Can only renounce role for self');
+            _revoke_role(role, account);
+        }
+    }
+
+    #[constructor]
+    fn constructor(admin: ContractAddress) {
+        initializer();
+        _grant_role(DEFAULT_ADMIN_ROLE, admin);
```
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/access/accesscontrol.cairo:68]: Yeah, we should remove it. I wanted to give users an example of how to set a constructor up, but that can be for documentation and presets. It doesn't belong in the actual mod

File: src/openzeppelin/tests/test_accesscontrol.cairo [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+fn test_initializer() {
+    AccessControl::initializer();
+    assert(AccessControl::supports_interface(IACCESSCONTROL_ID), 'Should support own interface');
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_constructor() {
+    AccessControl::constructor(ACCOUNT1());
+    assert(AccessControl::has_role(DEFAULT_ADMIN_ROLE, ACCOUNT1()), 'Admin role should be set');
```
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/tests/test_accesscontrol.cairo:31]: Agreed, we can include an AccessControl constructor (and tests) when we get to presets

File: src/openzeppelin/access/accesscontrol.cairo [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+#[contract]
+mod AccessControl {
+    // OZ modules
+    use super::IAccessControl;
+    use openzeppelin::introspection::erc165::ERC165;
+
+    // Constants
+    use openzeppelin::utils::constants::{DEFAULT_ADMIN_ROLE, IACCESSCONTROL_ID};
```
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/access/accesscontrol.cairo:19]: If I understand the question correctly, I separated the imports when we had to import each constant with the full path. It's irrelevant now, I'll consolidate and remove the `// Constants` comment

File: src/openzeppelin/tests/test_accesscontrol.cairo [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    contract_address_const::<2>()
+}
+
+fn setup() {
+    set_caller_address(ACCOUNT1());
+    AccessControl::constructor(ACCOUNT1());
+}
+
+#[test]
+#[available_gas(2000000)]
```
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/tests/test_accesscontrol.cairo:34]: Definitely not opposed! I'll look into it

File: src/openzeppelin/access/accesscontrol.cairo (hunk 11 above)
[the remark this answers is not in the document]
Comment (martriay) [reply src/openzeppelin/access/accesscontrol.cairo:16]: i meant collapsing imports in a single line!

File: src/openzeppelin/access/accesscontrol.cairo (hunk 7 above)
[the remark this answers is not in the document]
Comment (martriay) [reply src/openzeppelin/access/accesscontrol.cairo]: i think either way is fine, although we should settle on this since changing it later might be a breaking change. @ericnordelo do you have any preferences between having `ACCOUNT_ID` and `DEFAULT_ADMIN_ROLE` as constants in the `constants` module vs in their respective modules e.g. `openzeppelin::account::ACCOUNT_ID` and `openzeppelin::access::access_control::DEFAULT_ADMIN_ROLE`? i think i lean towards the latter

File: src/openzeppelin/access/accesscontrol.cairo (hunk 11 above)
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/access/accesscontrol.cairo:16]: Hahaha my bad! Not due to linter. To me, it seems unnecessary to write out entire paths for imports when they're coming from the source, no? Do you think it's more clear to write each one out?

```
use openzeppelin::utils::constants::{DEFAULT_ADMIN_ROLE, IACCESSCONTROL_ID};

vs

use openzeppelin::utils::constants::DEFAULT_ADMIN_ROLE;
use openzeppelin::utils::constants::IACCESSCONTROL_ID;
```

File: src/openzeppelin/tests/test_accesscontrol.cairo [hunk 13]
```diff
@@ -2,124 +2,308 @@ use openzeppelin::access::accesscontrol::AccessControl;
...
 }
 
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected: ('Caller is missing role', ))]
+fn test_assert_only_role_unauthorized_2() {
+    setup();
+    set_caller_address(AUTHORIZED());
+    AccessControl::assert_only_role(OTHER_ROLE);
+}
```
Comment (martriay) [inline src/openzeppelin/tests/test_accesscontrol.cairo:83]: what was the rationale to add this extra test?

File: src/openzeppelin/tests/test_accesscontrol.cairo [hunk 14]
```diff
@@ -2,124 +2,308 @@ use openzeppelin::access::accesscontrol::AccessControl;
...
+    assert(AccessControl::has_role(ROLE, AUTHORIZED()), 'Role should still be granted');
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_grant_role_already_granted() {
+    setup();
+    AccessControl::grant_role(ROLE, AUTHORIZED());
+    AccessControl::grant_role(ROLE, AUTHORIZED());
+    assert(AccessControl::has_role(ROLE, AUTHORIZED()), 'Role should still be held');
```
Comment (martriay) [inline src/openzeppelin/tests/test_accesscontrol.cairo:112]: how is this test different from the one above? i think i get the difference is that the above intends to check that it doesn't revert/passes, while the one below intends to check that the account still has the role. but are they really different?

File: src/openzeppelin/tests/test_accesscontrol.cairo [hunk 15]
```diff
@@ -2,124 +2,308 @@ use openzeppelin::access::accesscontrol::AccessControl;
...
+//
+// revoke_role
+//
+
+#[test]
+#[available_gas(2000000)]
+fn test_revoke_role_for_role_not_granted() {
+    setup();
+    AccessControl::revoke_role(ROLE, AUTHORIZED());
+    assert(!AccessControl::has_role(ROLE, AUTHORIZED()), 'Role should be revoked');
```
Comment (martriay) [inline src/openzeppelin/tests/test_accesscontrol.cairo:133]: we don't need this assert, don't we? we're just checking that revoking a non-granted role passes, no?

File: src/openzeppelin/tests/test_accesscontrol.cairo [hunk 16]
```diff
@@ -2,124 +2,308 @@ use openzeppelin::access::accesscontrol::AccessControl;
...
+//
+
+#[test]
+#[available_gas(2000000)]
+fn test_renounce_role_for_role_not_granted() {
+    setup();
+    set_caller_address(AUTHORIZED());
+
+    AccessControl::renounce_role(ROLE, AUTHORIZED());
+    assert(!AccessControl::has_role(ROLE, AUTHORIZED()), 'Role should be renounced');
```
Comment (martriay) [inline src/openzeppelin/tests/test_accesscontrol.cairo:177]:
```suggestion
    assert(!AccessControl::has_role(ROLE, AUTHORIZED()), 'Role should not be held');
```
or something similar. it might confuse since the intent is not to *renounce*, but to not revert. as the comment above, we might not even need the assert.

Comment (martriay) [reply src/openzeppelin/tests/test_accesscontrol.cairo:177]: we should either assume the role is not granted (i lean towards this), or at most assert initial state validity by checking the role is not granted **before** calling`renounce_role`. what does the assert after add?

currently it seems we're testing that `renounce_role` does not grant the role

File: src/openzeppelin/tests/test_accesscontrol.cairo [hunk 17]
```diff
@@ -2,124 +2,308 @@ use openzeppelin::access::accesscontrol::AccessControl;
...
 #[test]
 #[available_gas(2000000)]
 #[should_panic(expected: ('Can only renounce role for self', ))]
 fn test_renounce_role_unauthorized() {
     setup();
-    set_caller_address(ACCOUNT2());
-    AccessControl::renounce_role(DEFAULT_ADMIN_ROLE, ACCOUNT1());
+    AccessControl::grant_role(ROLE, AUTHORIZED());
+
+    // Admin is unauthorized caller
```
Comment (martriay) [inline src/openzeppelin/tests/test_accesscontrol.cairo:210]: good comment

File: src/openzeppelin/tests/test_accesscontrol.cairo [hunk 18]
```diff
@@ -2,124 +2,308 @@ use openzeppelin::access::accesscontrol::AccessControl;
...
-        'Should set the admin role'
+        AccessControl::get_role_admin(ROLE) == DEFAULT_ADMIN_ROLE, 'ROLE admin default should be 0'
     );
+    AccessControl::_set_role_admin(ROLE, OTHER_ROLE);
+    assert(AccessControl::get_role_admin(ROLE) == OTHER_ROLE, 'ROLE admin should be OTHER_ROLE');
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_change_an_admin_roles_admin() {
```
Comment (martriay) [inline src/openzeppelin/tests/test_accesscontrol.cairo:231]: i don't think i understand this test. what does it test?

**Review by martriay — COMMENTED** [review]
great improvements! the test suit looks very solid now. left some final suggestions.

File: src/openzeppelin/access/accesscontrol.cairo (hunk 7 above)
[the remark this answers is not in the document]
Comment (ericnordelo) [reply src/openzeppelin/access/accesscontrol.cairo]: For me also both ways are fine as long as we keep consistency, but I'm leaning towards the latter as well (respective modules), mostly because I think you get more info regarding these constants this way just by looking at the imports.

File: src/openzeppelin/access/accesscontrol.cairo (hunk 11 above)
[the remark this answers is not in the document]
Comment (martriay) [reply src/openzeppelin/access/accesscontrol.cairo:16]: i actually do

File: src/openzeppelin/access/accesscontrol.cairo [hunk 19]
```diff
@@ -11,41 +11,49 @@ trait IAccessControl {
...
     struct Storage {
         role_admin: LegacyMap<felt252, felt252>,
-        role_member: LegacyMap<(felt252, ContractAddress), bool>,
+        role_members: LegacyMap<(felt252, ContractAddress), bool>,
     }
 
+    /// Emitted when `account` is granted `role`.
+    ///
+    /// `sender` is the account that originated the contract call, an admin role
+    /// bearer (except when `_grant_role` is called during initialization from the constructor).
```
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/access/accesscontrol.cairo:27]: Hmm maybe something like:

```
    /// `sender` is the account that originated the contract call, an admin role
    /// bearer (except if `_grant_role` is called during initialization from the
    /// implementing contract's constructor).
```

And when we have the docsite updated, we can link setup instructions?

File: src/openzeppelin/tests/test_accesscontrol.cairo (hunk 13 above)
Comment (andrew-fleming) [reply src/openzeppelin/tests/test_accesscontrol.cairo:83]: This was to test that granting ROLE to AUTHORIZED won't allow AUTHORIZED to pass as other roles. It was missing the grant `ROLE` to `AUTHORIZED`

File: src/openzeppelin/tests/test_accesscontrol.cairo (hunk 14 above)
Comment (andrew-fleming) [reply src/openzeppelin/tests/test_accesscontrol.cairo:112]: Haha yeah, you're right. One should work

File: src/openzeppelin/tests/test_accesscontrol.cairo (hunk 16 above)
Comment (andrew-fleming) [reply src/openzeppelin/tests/test_accesscontrol.cairo:177]: Agreed in assuming the role is not granted and the assertion is not needed

File: src/openzeppelin/tests/test_accesscontrol.cairo (hunk 18 above)
Comment (andrew-fleming) [reply src/openzeppelin/tests/test_accesscontrol.cairo:231]: Hm it's really repeating the initial `test__set_role_admin`. I'll remove

File: src/openzeppelin/access/accesscontrol.cairo (hunk 19 above)
[the remark this answers is not in the document]
Comment (martriay) [reply src/openzeppelin/access/accesscontrol.cairo]: i don't think it changes much. +1 to linking setup instructions when we have docs.

[the remark this answers is not in the document]
Comment (martriay) [reply src/openzeppelin/access/accesscontrol.cairo]: i don't think it changes much. +1 to linking setup instructions when we have docs.

File: src/openzeppelin/tests/test_accesscontrol.cairo (hunk 13 above)
Comment (martriay) [reply src/openzeppelin/tests/test_accesscontrol.cairo]: let's make this clear in the test name then

File: src/openzeppelin/tests/test_accesscontrol.cairo [hunk 20]
```diff
@@ -60,7 +63,7 @@ fn test_has_role() {
 fn test_assert_only_role() {
     setup();
     AccessControl::grant_role(ROLE, AUTHORIZED());
-    set_caller_address(AUTHORIZED());
+    testing::set_caller_address(AUTHORIZED());
```
Comment (martriay) [inline src/openzeppelin/tests/test_accesscontrol.cairo:66]:
```suggestion
    testing::set_caller_address(AUTHORIZED);
```
we can now change these functions for top level constants, no?

**Review by martriay — COMMENTED** [review]
final lap! almost ready 💯

File: src/openzeppelin/tests/test_accesscontrol.cairo (hunk 20 above)
Comment (andrew-fleming) [reply src/openzeppelin/tests/test_accesscontrol.cairo]: Unfortunately, no. In the Account PR, the public key was a felt252 (which I think this is in reference to), so that can be a constant. No support yet for contract addresses though

**Review by martriay — APPROVED** [review]
fix conflicts and good to go!
